# Final Project. SQL

## Table of Contents

1. [Introduction](#intro)
2. [General Lookup](#general)
    * [Conclusion](#concl)
3. [Queries](#query)
    * [Conclusion](#concl1)
4. [Overall conclusion](#overall)


<a id="intro"></a>
## Introduction

<b>Project description</b>:
<br>The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

We've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

<b>Project structure</b>:
1. General information. Studying the given tables.
2. Data queries. Making calculations for the proposition. 
3. Conclusion.

<a id="general"></a>
## General Lookup

In [1]:
!pip install -Uq psycopg2-binary

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<b>books</b>

In [3]:
query="""
SELECT *
FROM books 

LIMIT 10"""
table = pd.io.sql.read_sql(query, con = engine)
table

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [5]:
query="""
SELECT MIN(num_pages) AS min_pages, 
       MAX(num_pages) AS max_pages,
       MIN(publication_date) AS first_date,
       MAX(publication_date) AS last_date,
       COUNT(DISTINCT book_id) AS num_books
FROM books 
"""
values = pd.io.sql.read_sql(query, con = engine)
values

min_pages  max_pages  first_date   last_date  num_books
0         14       2690  1952-12-01  2020-03-31       1000

*Conclusion*: Our first table is `books`. This table contains variables:
* `book_id` - key variable for `ratings` and `reviews` tables
* `author_id` - key variable for `authors` table.
* `title` - title of the book. There are 1000 titles in the data.
* `num_pages` - number of pages in a book. We've seen that maximum is 2690 and minimum is 14.
* `publication_date` - first book in the table was published on 01.12.1952 and the newest book was published on 31.03.2020.
* `publisher_id` - key variable for `publishers` table.

<b>authors</b>

In [6]:
query="""
SELECT *
FROM authors 

LIMIT 10"""
table = pd.io.sql.read_sql(query, con = engine)
table

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [7]:
query="""
SELECT COUNT(DISTINCT author_id)
FROM authors
"""

temp = pd.io.sql.read_sql(query, con = engine)
temp

count
0    636

*Conclusion*: this is a small table with only 2 fields.
* `author_id`
* `author` - full name of the author. There are 636 authors in our data.

<b>ratings</b>

In [8]:
query="""
SELECT *
FROM ratings 

LIMIT 10"""
table = pd.io.sql.read_sql(query, con = engine)
table

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [9]:
query="""
SELECT MIN(rating) AS min_rating,
       MAX(rating) AS max_rating
FROM ratings 

LIMIT 10"""
values = pd.io.sql.read_sql(query, con = engine)
values

min_rating  max_rating
0           1           5

*Conclusion*:this table contains 4 fields.
* `rating_id`
* `book_id` - key variable for `books` table.
* `username`
* `rating` - rating system goes from 1 to 5.

<b>reviews</b>

In [10]:
query="""
SELECT *
FROM reviews

LIMIT 10"""
table = pd.io.sql.read_sql(query, con = engine)
table

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

*Conclusion*: this table also contains 4 fields.
* `review_id`
* `book_id` - key variable for `books` table.
* `username`
* `text` - the text of the review

<b>publishers</b>

In [11]:
query="""
SELECT *
FROM publishers

LIMIT 10"""
table = pd.io.sql.read_sql(query, con = engine)
table

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [12]:
query="""
SELECT COUNT(DISTINCT publisher_id)
FROM publishers
"""
n_rows = pd.io.sql.read_sql(query, con = engine)
n_rows

count
0    340

*Conclusion*: small table containing only 2 fields.
* `publisher_id`
* `publisher` - the name of the publisher. There are 340 publishers in the data.

<a id="concl"></a>
### Conclusion

We have a total of 5 relatively small tables containing information on books, authors, publishers, reviews and ratings.

* `books`, `reviews` and `ratings` have shared key variable `book_id`.
* `books` and `authors` share `author_id` variable.
* `books` and `publishers` - `publisher_id` variable.

We have data on total of 1000 books written by 636 authors and published by 340 publishers. The maximum number of pages is 2690 while minimum is 14 (perhaps it is something like a brochure). The latest published book was published on 31.03.20.

<a id="query"></a>
## Table queries

<b>Number of books released after January 1, 2000.</b>

In [13]:
query="""
SELECT COUNT(DISTINCT book_id)
FROM books 
WHERE CAST(publication_date AS timestamp) > '2000-01-01'
"""
num_books = pd.io.sql.read_sql(query, con = engine)
num_books

count
0    819

*Conclusion*: out of 1000 books in the data 819 were released after 01.01.2000, meaning that 80% of the books we have in the data were released in last 20 years.

<b>The number of user reviews and the average rating for each book.</b>

In [14]:
query="""
SELECT title,
       COUNT(review_id) AS num_reviews, 
       AVG(rating) AS avg_rating
FROM books AS b
LEFT JOIN ratings AS rat USING(book_id)
LEFT JOIN reviews AS rev USING(book_id)
GROUP BY title
ORDER BY num_reviews DESC

"""

values_per_book = pd.io.sql.read_sql(query, con = engine)
values_per_book.head(10)

title  num_reviews  avg_rating
0                            Twilight (Twilight  #1)         1120    3.662500
1                The Hobbit  or There and Back Again          528    4.125000
2                             The Catcher in the Rye          516    3.825581
3  Harry Potter and the Prisoner of Azkaban (Harr...          492    4.414634
4  Harry Potter and the Chamber of Secrets (Harry...          480    4.287500
5               Angels & Demons (Robert Langdon  #1)          420    3.678571
6  Harry Potter and the Order of the Phoenix (Har...          375    4.186667
7  The Lightning Thief (Percy Jackson and the Oly...          372    4.080645
8  The Fellowship of the Ring (The Lord of the Ri...          370    4.391892
9                                        Animal Farm          370    3.729730

*Conclusion*: with this query we've created a reference table with number of reviews and average rating for each book. Number of reviews can tell us about the popularity of the book and average rating about it's quality in the eyes of the reader.

After ordering by the number of reviews we see that not always most popular means the highest quality: `Twilight` title gathered twice number of reviews compared to second place `Hobbit`, while having pretty low average rating - 3.7.

In this industry the most profitable book can be the one bringing up the most controversy: there are reader who absolutley love it and those who absolutely hate it, which brings up discussions and desire to defend their opinion.

<b>Publisher that has released the greatest number of books with more than 50 pages (this will help to exclude brochures and similar publications).</b>

In [15]:
query="""
SELECT publisher, 
       COUNT (book_id) AS num_books
FROM books AS b
LEFT JOIN publishers AS pub USING(publisher_id)
WHERE num_pages > 50
GROUP BY publisher
ORDER BY num_books DESC
LIMIT 5
"""

greatest_publisher = pd.io.sql.read_sql(query, con = engine)
greatest_publisher

publisher  num_books
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25
3          Penguin Classics         24
4                    Bantam         19

*Conclusion*: `Penguin Books` is the most active publisher, having released 42 books from our table. There are total of 340 publishers in the data, so 42 is a pretty big number for 1000 books that we have.

For comparison we also showed other top4 publishers, and number 2 is `Vintage` with 31 books. Other publishers don't even reach 30 books.

<b>The author with the highest average book rating (only for books with at least 50 ratings).</b>

In [16]:
query="""
SELECT author, AVG(rating) AS avg_rating
FROM books AS b
LEFT JOIN authors USING(author_id)
LEFT JOIN ratings USING(book_id)
WHERE b.book_id IN(
    SELECT book_id
    FROM books
    LEFT JOIN ratings AS r USING(book_id)
    GROUP BY book_id
    HAVING COUNT(r.rating_id) >= 50
)
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
"""

greatest_author = pd.io.sql.read_sql(query, con = engine)
greatest_author

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

*Conclusion*: the author with highest average rating `4.3` is <b>J.K. Rowling</b> - the author of Harry Potter series, loved by children and adults. We also saw previously that several of her books are in top 10 of books with most numerous reviews.

There's no denying `Harry Potter` conquered the hearts of generations, these books will sell well for years to come. Additionally, everything Rowling writes will undoubtedly attract a lot of attention.

<b>The average number of text reviews among users who rated more than 50 books. </b>

In [17]:
query = """
SELECT AVG(t.n_reviews)
FROM (

SELECT COUNT(review_id) AS n_reviews
FROM reviews
WHERE username IN(
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 50
    )
GROUP BY username
) AS t
"""

avg_Nreviews = pd.io.sql.read_sql(query, con=engine)
avg_Nreviews

avg
0  24.333333

*Conclusion*: readers that actively rate books, 50 at least, leave half of this number text reviews. It is logical, as it is much easier to give a book rating from 1 to 5 than to write a whole review. 

<a id="concl2"></a>
## Overall Conclusion

Concluding our research we can say that our database contains data on 1000 books, 819 of them, or about 80% were released in last 20 years.

We've created a reference table with number of reviews and average rating of each book, which shows book's popularity and quality in reader's eyes. We recommend to pay attention to the fact that books with highest number of reviews don't always have highest average rating (e.g. `Twilight` with 1120 reviews and only 3.7 average rating).

We've found that the most active publisher is `Penguin Books`, having released 42 books with more than 50 pages from our dataset. And the author with highest average rating turned out to be <b>J.K. Rowling</b>, author of *'Harry Potter'* series. Chances are that any book she releases will attract a lot of attention.

At last, we've found that even the most active readers who've rated at least 50 books, have written about 24 text reviews on average. Meaning that there's a pretty small chance that an active user will write a text rerview for a book. For regular readers chances are even slimmer.